### ISSUE TITLE
Kernel debugging sometimes freezes

### ISSUE THREAD

**Issue by gendx** [tock/tock#1629]
While debugging OpenSK, I discovered the following bug (https://github.com/google/OpenSK/issues/58). I'm filing an issue here as well as it's likely on Tock OS side (although unclear which part).

My intuition is that when writing to the console output, Tock is supposed to trigger some alarm to indicate that a buffer was written. But in some cases the alarm doesn't fire, and subsequent attempts to write to the console (via `debug!` or the `Console` driver) then just fail or block.

Although I could observe and reproduce this on a specific configuration, it's unclear to me whether the issue is specific to Segger RTT, the nRF52840-DK, or is a more general bug in `kernel/debug`, `kernel/hil/uart` or `capsules/console` (or some related component).

## Expected output

The expected output when starting [OpenSK](https://github.com/google/OpenSK) is as follows:

```
Initialization complete. Entering main loop
NRF52 HW INFO: Variant: AAD0, Part: N52840, Package: QI, Ram: K256, Flash: K1024
usbc::start() - State=PoweredOn
Your chip is NRF52840 revision 3. Although this USB implementation should be compatible, your chip hasn't been tested.
usbc::start() - subscribing to interrupts.
enable_out_endpoint_(0) - State=Started
enable_in_out_endpoint_(1) - State=Started
attach() - State=Started
enable_pullup() - State=Started
Enabling USB pullups
New state is Attached
usbc::ready not implemented
usbc::suspend not implemented
usbc::resume not implemented
Bus reset
Bus reset
Set Address = 8
```

## Actual behavior

Sometimes the console output stops at the following line. That is, `debug!` statements in the kernel don't appear in console output.

```
Initialization complete. Entering main loop
NRF52 HW INFO: Variant: AAD0, Part: N52840, Package: QI, Ram: K256, Flash: K1024
usbc::start() - State=PoweredOn
Your chip is NRF52840 revision 3. Although this USB implementation should be compatible, your chip hasn't been tested.
usbc::start() - subscribing to interrupts.
enable_out_endpoint_(0) - State=Started
enable_in_out_endpoint_(1) - State=Started
attach() - State=Started
enable_pullup() - State=Started
Enabling USB pullups
New state is Attached
usbc::ready not implemented
```

Although it doesn't freeze every time, it seems to always freeze after `usbc::ready not implemented`. The relevant `debug!` statements are here:

https://github.com/tock/tock/blob/3a7d6b775d972798bfd731cba8365b58fab27175/chips/nrf52/src/usbd.rs#L1470-L1485

where `internal_warn!` just calls `debug!`:

https://github.com/tock/tock/blob/3a7d6b775d972798bfd731cba8365b58fab27175/chips/nrf52/src/usbd.rs#L44-L48

Additionally, it appears that the `Console` driver freezes as well if used after such a freeze in the kernel. In other words, no console callback is delivered to applications.

## Platform information

I observed this bug with the following configuration.

- Platform: nRF52840-DK
- Debug output: both [Segger RTT](https://github.com/tock/tock/blob/master/capsules/src/segger_rtt.rs) (see https://github.com/tock/tock/pull/1556) and UART.

However, it's unclear to me whether this bug is specific to this configuration or rather a more generic bug in the kernel's handling of `debug!`.

**Comment by gendx** [comment]
I added a few `debug_gpio` statements in `kernel/src/debug.rs`, and in the "frozen" case it appears that after a `transmit_buffer` call here:

https://github.com/tock/tock/blob/2d88744ec83c589be75fe7106dec73da27c6e847/kernel/src/debug.rs#L307

the `transmitted_buffer` callback never arrives here:

https://github.com/tock/tock/blob/2d88744ec83c589be75fe7106dec73da27c6e847/kernel/src/debug.rs#L319-L328

**Comment by gendx** [comment]
Updated the platform information. I observed it regardless of the debug output (Segger RTT or UART).

**Comment by gendx** [comment]
I think I've found the culprit.

- In `DynamicDeferredCall::call_while`, a `call_pending` is first set to `false`.
- Then the function goes on to loop over scheduled callbacks, but only while some external predicate holds (`.take_while(|_| f())`). This means that this predicate can disrupt the loop, in which case no callback is called.
- But because `call_pending` was set to `false`, the next time `call_while` is called, everything gets skipped.

https://github.com/tock/tock/blob/83c65e7742513693606e23661a1673112ed3e9f1/kernel/src/common/dynamic_deferred_call.rs#L227-L248

Removing the `call_pending` sentinel indeed removes the issue.

---

Now, how does the race condition with `take_while` happens? The predicate is that the chip doesn't have any interrupt pending:

https://github.com/tock/tock/blob/f7f3b420edd1949d9a0d49811d119d0b72c64f2d/kernel/src/sched.rs#L216

My guess is that in the scenario I observed, a bunch of USB interrupts occur in a small amount of time, so that between the time the first `debug!` statement is issued (following an USB interrupt!) and the time the `transmitted_buffer` callback is performed, the other USB interrupt occurs.

All-in-all, `call_pending` is subject to a race condition w.r.t. interrupts.
